File: python/BestTimetoBuyandSellStockIII.py
class Solution:
    # @param {integer[]} prices
    # @return {integer}
    
##    def maxProfit(self, prices):
##        if not prices:
##            return 0
##        maxpro1, minind, maxind = self.auxMaxProfit(prices)
##        # can sell and buy at the same time
##        # if the first sell and last buy is between minind, maxind
##        # we can always adapt the first buy and last sell to minind, maxind
##        maxpro2 = max(self.auxMaxProfit(prices[:minind+1])[0],
##                     self.auxMaxProfit(prices[maxind:])[0],
##                     self.auxMaxProfit(prices[minind:maxind+1][::-1])[0])
##        maxpro = maxpro1 + maxpro2
##        return maxpro
##    
##    def auxMaxProfit(self, prices):
##        if not prices:
##            return 0, 0, 0
##        maxpro = 0
##        curmin = prices[0]
##        curminind = 0
##        minind = 0
##        maxind = 0
##        for i in range(len(prices)):
##            if prices[i]<curmin:
##                curmin = prices[i]
##                curminind = i
##            if prices[i]-curmin > maxpro:
##                maxpro = prices[i]-curmin
##                maxind = i
##                # notice minind and curminind
##                minind = curminind
##        return maxpro, minind, maxind
    def maxProfit(self, prices):
        hold1 = -2**31
        hold2 = -2**31
        release1 = 0
        release2 = 0
        for i in prices:
            release2 = max(release2, hold2 + i)
            hold2 = max(hold2, release1 - i)
            release1 = max(release1, hold1 + i)
            hold1 = max(hold1, -i)
        return release2

import random
prices = [random.randint(1,10) for i in range(10)]
print(prices)
print(Solution().maxProfit(prices))

File: cpp/Best Time to Buy and Sell Stock III.cpp
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;

class Solution {
public:
	int maxProfit(vector<int>& prices) {
		if (prices.size() < 2) return 0;

		const int n = prices.size();
		vector<int> f(n, 0);
		vector<int> g(n, 0);

		for (int i = 1, valley = prices[0]; i < n; ++i) {
			valley = min(valley, prices[i]);
			f[i] = max(f[i-1], prices[i] - valley);
		}

		for (int i = n - 2, peak = prices[n - 1]; i >= 0; --i) {
			peak = max(peak, prices[i]);
			g[i] = max(g[i], peak - prices[i]);
		}

		int max_profit = 0;
		for (int i = 0; i < n; ++i) 
			max_profit = max(max_profit, f[i] + g[i]);

		return max_profit;
	}
};

int main() {
	vector<int> prices;
	cout<<Solution().maxProfit(prices)<<endl;
	system("pause");
	return 0;
}